thirdparty/patches/opencensus-cpp-shutdown-api.patch

```diff
@@ -24,49 +100,43 @@ diff --git opencensus/stats/internal/stats_exporter.cc b/opencensus/stats/intern
  
  void StatsExporterImpl::RunWorkerLoop() {
    absl::Time next_export_time = GetNextExportTime();
--  while (true) {
-+  bool thread_started = false;
-+  {
-+    absl::MutexLock l(&mu_);
-+    bool thread_started = thread_started_;
-+  }
-+  while (thread_started) {
-     // SleepFor() returns immediately when given a negative duration.
-     absl::SleepFor(next_export_time - absl::Now());
-     // In case the last export took longer than the export interval, we
-     // calculate the next time from now.
-     next_export_time = GetNextExportTime();
-     Export();
+   while (true) {
 +    {
 +      absl::MutexLock l(&mu_);
-+      thread_started = thread_started_;
++      if (!thread_started_) {
++        break;
++      }
 +    }
-   }
- }
- 
+     // SleepFor() returns immediately when given a negative duration.
+     absl::SleepFor(next_export_time - absl::Now());
+     // In case the last export took longer than the export interval, we
+@@ -114,6 +134,12 @@ void StatsExporterImpl::RunWorkerLoop() {
  // StatsExporter
  // -------------
  
 +void StatsExporter::Shutdown() {
 +  StatsExporterImpl::Get()->Shutdown();
++  StatsExporter::ExportForTesting();
```

Comment:
Why we need do exporting in shutdown? we can't guarantee all metric data will be flushed.

Reply:
Ah, good point. Maybe I should add flush & export for testing in ray shutdown to make sure all remaining metrics are exported?